drivers/pinmux/pinmux_ite_it8xxx2.c

```diff
@@ -13,76 +13,159 @@
+
+	/* Common settings for alternate function. */
+	*reg_gpcr &= ~(GPCR_PORT_PIN_MODE_INPUT |
+		GPCR_PORT_PIN_MODE_OUTPUT);
+
+	switch (func) {
+	case IT8XXX2_PINMUX_FUNC_1:
+		/* Func1: Alternate function has been set above. */
+		break;
+	case IT8XXX2_PINMUX_FUNC_2:
+		/* Func2: WUI function: turn the pin into an input */
+		*reg_gpcr = (*reg_gpcr | GPCR_PORT_PIN_MODE_INPUT) &
+			~GPCR_PORT_PIN_MODE_OUTPUT;
+		break;
+	case IT8XXX2_PINMUX_FUNC_3:
+		/*
+		 * Func3: In addition to the alternate setting above,
+		 *        Func3 also need to set the general control.
+		 */
+		*reg_func3_gcr |= pinmux_config->func3_en_mask[pin];
+		break;
+	case IT8XXX2_PINMUX_FUNC_4:
+		/*
+		 * Func4: In addition to the alternate setting above,
+		 *        Func4 also need to set the general control.
+		 */
+		*reg_func4_gcr |= pinmux_config->func4_en_mask[pin];
+		break;
+	default:
+		LOG_ERR("This function is not supported");
+		return -EINVAL;
 	}
 
 	return 0;
 }
 
 static int pinmux_it8xxx2_get(const struct device *dev,
-		uint32_t pin, uint32_t *func)
+			      uint32_t pin, uint32_t *func)
 {
-	const struct pinmux_it8xxx2_config *config = DEV_CFG(dev);
-	uint32_t reg;
-	uint8_t val;
-	int g, i;
+	const struct pinmux_it8xxx2_config *pinmux_config = DEV_CFG(dev);
+
+	volatile uint8_t *reg_gpcr =
+		(uint8_t *)(pinmux_config->reg_gpcr + pin);
 
 	if (pin >= IT8XXX2_PINMUX_PINS || func == NULL) {
 		return -EINVAL;
 	}
-	g =  pin / PIN_REG_OFFSET;
-	i =  pin % PIN_REG_OFFSET;
-	reg = config->base + g;
-	val = sys_read8(reg);
-	*func = (val & (IT8XXX2_PINMUX_IOF1 << pin)) ?
-		IT8XXX2_PINMUX_IOF1 : IT8XXX2_PINMUX_IOF0;
+
+	*func = (*reg_gpcr & (GPCR_PORT_PIN_MODE_INPUT |
+		GPCR_PORT_PIN_MODE_OUTPUT)) ?
+		IT8XXX2_PINMUX_FUNC_2 : IT8XXX2_PINMUX_FUNC_1;
+
+	/* TODO: IT8XXX2_PINMUX_FUNC_3 & IT8XXX2_PINMUX_FUNC_4 */
 
 	return 0;
 }
 
 static int pinmux_it8xxx2_pullup(const struct device *dev,
-		uint32_t pin, uint8_t func)
+				 uint32_t pin, uint8_t func)
 {
+	const struct pinmux_it8xxx2_config *pinmux_config = DEV_CFG(dev);
+
+	volatile uint8_t *reg_gpcr =
+		(uint8_t *)(pinmux_config->reg_gpcr + pin);
+
+	if (func == PINMUX_PULLUP_ENABLE) {
+		*reg_gpcr = (*reg_gpcr | GPCR_PORT_PIN_MODE_PULLUP) &
+			~GPCR_PORT_PIN_MODE_PULLDOWN;
+	} else { /* No pull up/down */
+		*reg_gpcr &= ~(GPCR_PORT_PIN_MODE_PULLUP |
+			GPCR_PORT_PIN_MODE_PULLDOWN);
+	}
+
 	return -ENOTSUP;
 }
 
 static int pinmux_it8xxx2_input(const struct device *dev,
-		uint32_t pin, uint8_t func)
+				uint32_t pin, uint8_t func)
 {
-	return -ENOTSUP;
+	const struct pinmux_it8xxx2_config *pinmux_config = DEV_CFG(dev);
+
+	volatile uint8_t *reg_gpcr =
+		(uint8_t *)(pinmux_config->reg_gpcr + pin);
+
+	if (func == PINMUX_INPUT_ENABLED) {
+		*reg_gpcr = (*reg_gpcr | GPCR_PORT_PIN_MODE_INPUT) &
+			~GPCR_PORT_PIN_MODE_OUTPUT;
+	} else { /* func == PINMUX_OUTPUT_ENABLED */
+		*reg_gpcr = (*reg_gpcr | GPCR_PORT_PIN_MODE_OUTPUT) &
+			~GPCR_PORT_PIN_MODE_INPUT;
+	}
```

Comment:
Change to verify func==OUTPUT, instead of making it the default.  I also suggest reading the gpcr register once.

gpcr = *reg_gpcr & ~(MODE_INPUT | MODE_OUTPUT);
if (func == INPUT)
    *reg_gpcr = gpcr | MODE_INPUT;
else if (func == OUTPUT)
    *reg_gpcr = gpcr | MODE_OUTPUT:
else
    return -EINVAL;

Reply:
Done